File: geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-api/src/main/java/org/locationtech/geomesa/fs/storage/api/PartitionScheme.java
/***********************************************************************
 * Copyright (c) 2013-2019 Commonwealth Computer Research, Inc.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Apache License, Version 2.0
 * which accompanies this distribution and is available at
 * http://www.opensource.org/licenses/apache2.0.php.
 ***********************************************************************/

package org.locationtech.geomesa.fs.storage.api;

import org.opengis.feature.simple.SimpleFeature;
import org.opengis.filter.Filter;

import java.util.List;
import java.util.Map;
import java.util.Optional;

public interface PartitionScheme {

    /**
     * Name of this partition scheme
     *
     * @return name
     */
    String getName();

    /**
     * Return the partition in which a SimpleFeature should be stored
     *
     * @param feature simple feature
     * @return partition name
     */
    String getPartition(SimpleFeature feature);

    @Deprecated
    default List<String> getPartitions(Filter filter) {
        // default implementation provides API compatibility but will throw an error if invoked
        throw new AbstractMethodError();
    }

    /**
     * Return a list of modified filters and partitions. Each filter will have been simplified to
     * remove any predicates that are implicitly true for the associated partitions
     *
     * If the filter does not constrain partitions at all, then an empty option will be returned,
     * indicating all partitions must be searched. If the filter excludes all potential partitions,
     * then an empty list of partitions will be returned
     *
     * Note that this operation is based solely on the partition scheme, so may return partitions
     * that do not actually exist in a given storage instance
     *
     * @param filter filter
     * @return list of simplified filters and partitions
     */
    default Optional<List<FilterPartitions>> getFilterPartitions(Filter filter) {
        return Optional.empty();
    }

    /**
     *
     * @return the max depth this partition scheme goes to
     */
    int getMaxDepth();

    /**
     * Are partitions stored as leaves (multiple partitions in a single folder), or does each
     * partition have a unique folder. Using leaf storage can reduce the level of nesting and make
     * file system operations faster in some cases.
     *
     * @return leaf
     */
    boolean isLeafStorage();

    /**
     * Options used to configure this scheme - @see PartitionSchemeFactory
     *
     * @return options
     */
    Map<String, String> getOptions();
}

File: geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-api/src/main/scala/org/locationtech/geomesa/fs/storage/api/PartitionScheme.scala
/***********************************************************************
 * Copyright (c) 2013-2025 General Atomics Integrated Intelligence, Inc.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Apache License, Version 2.0
 * which accompanies this distribution and is available at
 * https://www.apache.org/licenses/LICENSE-2.0
 ***********************************************************************/

package org.locationtech.geomesa.fs.storage.api

import com.typesafe.scalalogging.LazyLogging
import org.geotools.api.feature.simple.SimpleFeature
import org.geotools.api.filter.Filter
import org.locationtech.geomesa.fs.storage.api.PartitionScheme.PartitionRange
import org.locationtech.geomesa.fs.storage.api.PartitionScheme.RangeBuilder.BoundsOrdering
import org.locationtech.geomesa.fs.storage.api.StorageMetadata.PartitionKey

import scala.collection.mutable.ArrayBuffer

/**
  * Scheme for partitioning features into various named partitions (e.g. file paths) on disk, for
  * faster querying. Partition schemes do not have any persistent state, they only know how to map features
  * and filters to partition names
  */
trait PartitionScheme {

  /**
   * Name of this partition scheme
   *
   * @return
   */
  def name: String

  /**
    * Return the partition in which a SimpleFeature should be stored
    *
    * @param feature simple feature
    * @return partition name
    */
  def getPartition(feature: SimpleFeature): PartitionKey

  /**
   * Get partitions that intersect the given filter
   *
   * If the filter does not constrain partitions at all, then an empty option will be returned. If
   * the filter excludes all potential partitions, then an empty list will be returned
   *
   * @param filter filter
   * @return list of intersecting filters
   */
  def getRangesForFilter(filter: Filter): Option[Seq[PartitionRange]]

  /**
   * Enumerate all the partitions that intersect with the given filter
   *
   * If the filter does not constrain partitions at all, then an empty option will be returned. If
   * the filter excludes all potential partitions, then an empty list will be returned
   *
   * Note that this may return a large number of partitions if the filter is not very selective
   *
   * @param filter filter
   * @return
   */
  def getPartitionsForFilter(filter: Filter): Option[Seq[PartitionKey]]

  /**
   * Get a filter that will cover a partitions, i.e. the filter will return all features
   * in the given partition and none from other partitions
   *
   * @param partition partition to cover
   * @return filter
   */
  def getCoveringFilter(partition: PartitionKey): Filter
}

object PartitionScheme extends LazyLogging {

  /**
   * Ranged bounds
   *
   * @param name partition scheme name
   * @param lower lower bound, inclusive
   * @param upper upper bound, exclusive
   */
  case class PartitionRange(name: String, lower: String, upper: String) {

    /**
     * Is the value contained in this bounds
     *
     * @param value partition value
     * @return
     */
    def contains(value: String): Boolean = value >= lower && value < upper

    /**
     * Attempt to merge two bounds. Only overlapping bounds will result in a successful merge. Trying to merge
     * bounds from a different partition scheme is a logical error.
     *
     * @param other bounds to merge
     * @return
     */
    def merge(other: PartitionRange): Option[PartitionRange] = {
      if (lower <= other.lower) {
        if (upper >= other.upper) {
          Some(this)
        } else if (upper >= other.lower) {
          Some(PartitionRange(name, lower, other.upper))
        } else {
          None
        }
      } else if (lower > other.upper) {
        None
      } else if (upper >= other.upper) {
        Some(PartitionRange(name, other.lower, upper))
      } else {
        Some(other)
      }
    }
  }

  /**
   * Class to merge overlapping ranges.
   *
   * Our bounds extraction does not produce any overlapping ranges, but once converted to partitions there
   * may be some overlap.
   */
  class RangeBuilder {

    private val ranges = ArrayBuffer.empty[PartitionRange]

    def +=(range: PartitionRange): Unit = ranges += range

    def result(): Seq[PartitionRange] = {
      val all = ranges.sorted(BoundsOrdering)
      if (all.lengthCompare(1) <= 0) {
        all.toSeq
      } else {
        // merge any overlapping ranges that resulted
        val result = Seq.newBuilder[PartitionRange]
        var current = all.head
        all.tail.foreach { range =>
          current.merge(range) match {
            case None =>
              result += current
              current = range
            case Some(merged) =>
              current = merged
          }
        }
        result += current
        result.result()
      }
    }
  }

  object RangeBuilder {
    private val BoundsOrdering = Ordering.by[PartitionRange, String](_.lower)
  }
}
